File: android/app/src/main/java/com/metanav/app/frames/GlassesFrameSource.kt
package com.metanav.app.frames

import android.graphics.Bitmap
import android.util.Log
import com.meta.wearable.dat.camera.Camera
import com.meta.wearable.dat.camera.addCamera
import com.meta.wearable.dat.camera.types.StreamConfiguration
import com.meta.wearable.dat.camera.types.StreamState
import com.meta.wearable.dat.camera.types.VideoQuality
import com.meta.wearable.dat.core.Wearables
import com.meta.wearable.dat.core.selectors.DeviceSelector
import com.meta.wearable.dat.core.session.DeviceSession
import com.meta.wearable.dat.core.session.DeviceSessionState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeoutOrNull

/**
 * Streams the glasses' camera through the Meta Wearables Device Access Toolkit.
 * Lifecycle: create + start a [DeviceSession], wait for STARTED, attach a camera with an
 * uncompressed low-resolution stream, convert each I420 frame to a bitmap.
 */
class GlassesFrameSource(private val deviceSelector: DeviceSelector) : FrameSource {
    override val kind = SourceKind.GLASSES

    private val _state = MutableStateFlow<SourceState>(SourceState.Idle)
    override val state: StateFlow<SourceState> = _state.asStateFlow()

    private val _frames = MutableSharedFlow<RgbFrame>(replay = 0, extraBufferCapacity = 1)
    override val frames: Flow<RgbFrame> = _frames.asSharedFlow()

    private var scope: CoroutineScope? = null
    private var session: DeviceSession? = null
    private var camera: Camera? = null
    private var frameJob: Job? = null
    private var reuseBitmap: Bitmap? = null

    override suspend fun start() {
        if (session != null) return
        val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default).also { this.scope = it }
        _state.value = SourceState.Connecting

        val created = Wearables.createSession(deviceSelector).getOrElse { error ->
            fail("Could not start a glasses session: ${error.message ?: error}")
            return
        }
        session = created
        scope.launch {
            created.state.collect { s ->
                when (s) {
                    DeviceSessionState.PAUSED -> _state.value = SourceState.Paused
                    DeviceSessionState.STOPPED -> if (_state.value !is SourceState.Error) _state.value = SourceState.Idle
                    else -> Unit
                }
            }
        }
        scope.launch {
            created.errors.collect { error -> Log.w(TAG, "Session error: ${error.description}") }
        }
        created.start()

        val started = withTimeoutOrNull(20_000) {
            created.state.first { it == DeviceSessionState.STARTED || it == DeviceSessionState.STOPPED }
        }
        if (started != DeviceSessionState.STARTED) {
            fail(if (started == null) "Timed out connecting to the glasses. Are they on and unfolded?" else "The glasses session stopped before it started.")
            return
        }

        val config = StreamConfiguration(
            videoQuality = VideoQuality.LOW,
            frameRate = FRAME_RATE,
            compressVideo = false,
        )
        val cam = created.addCamera(config).getOrElse { error ->
            fail("Could not open the glasses camera: ${error.message ?: error}")
            return
        }
        camera = cam
        val stream = cam.stream
        scope.launch {
            stream.state.collect { s ->
                _state.value = when (s) {
                    StreamState.STREAMING -> SourceState.Streaming
                    StreamState.STARTING, StreamState.STARTED -> SourceState.WaitingForDevice
                    StreamState.PAUSED -> SourceState.Paused
                    StreamState.STOPPED, StreamState.CLOSED, StreamState.STOPPING ->
                        if (_state.value is SourceState.Error) _state.value else SourceState.Idle
                }
            }
        }
        scope.launch {
            stream.errorStream.collect { error -> Log.w(TAG, "Stream error: ${error.description}") }
        }
        frameJob = scope.launch {
            stream.videoStream.collect { frame ->
                if (frame.isCompressed || frame.isCodecConfig) return@collect
                val bmp = YuvConverter.i420ToBitmap(frame.buffer, frame.width, frame.height, null) ?: return@collect
                _frames.tryEmit(RgbFrame(bmp, System.currentTimeMillis()))
            }
        }
        stream.start().onFailure { error, _ ->
            fail("Could not start streaming: ${error.description}")
        }
    }

    override fun stop() {
        frameJob?.cancel(); frameJob = null
        try { camera?.stop() } catch (t: Throwable) { Log.w(TAG, "camera.stop failed", t) }
        camera = null
        try { session?.stop() } catch (t: Throwable) { Log.w(TAG, "session.stop failed", t) }
        session = null
        scope?.cancel(); scope = null
        reuseBitmap = null
        if (_state.value !is SourceState.Error) _state.value = SourceState.Idle
    }

    private fun fail(message: String) {
        Log.e(TAG, message)
        _state.value = SourceState.Error(message)
        stop()
        _state.value = SourceState.Error(message)
    }

    companion object {
        private const val TAG = "Metanav:Glasses"
        private const val FRAME_RATE = 15
    }
}

File: ios/Metanav/Frames/GlassesFrameSource.swift
import CoreMedia
import Foundation
import MWDATCamera
import MWDATCore
import UIKit

/// Streams the glasses' camera through the Meta Wearables Device Access Toolkit.
/// Lifecycle: create + start a `DeviceSession`, wait for `.started`, attach a camera with an
/// uncompressed low-resolution stream, forward each frame's pixel buffer.
final class GlassesFrameSource: FrameSource {
    let kind = SourceKind.glasses
    var onState: ((SourceState) -> Void)?
    var onFrame: ((CapturedFrame) -> Void)?

    private let wearables: WearablesInterface
    private let deviceSelector: DeviceSelector
    private var session: DeviceSession?
    private var camera: MWDATCamera.Camera?
    private let tokens = ListenerTokenBag()
    private var failed = false

    init(wearables: WearablesInterface, deviceSelector: DeviceSelector) {
        self.wearables = wearables
        self.deviceSelector = deviceSelector
    }

    func start() async {
        guard session == nil else { return }
        failed = false
        onState?(.connecting)

        let session: DeviceSession
        do {
            session = try wearables.createSession(deviceSelector: deviceSelector)
        } catch {
            fail("Could not start a glasses session: \(error.localizedDescription)")
            return
        }
        self.session = session

        session.statePublisher.listen { [weak self] state in
            guard let self else { return }
            switch state {
            case .paused: self.onState?(.paused)
            case .stopped: if !self.failed { self.onState?(.idle) }
            default: break
            }
        }.store(in: tokens)
        session.errorPublisher.listen { error in
            NSLog("[Metanav] glasses session error: \(error.description)")
        }.store(in: tokens)

        do {
            try session.start()
        } catch {
            fail("Could not start the glasses session: \(error.localizedDescription)")
            return
        }

        // Wait (bounded) for the session to come up before attaching the camera. Polling the
        // published state avoids depending on whether the state stream replays its current value.
        var waitedMs = 0
        while session.state != .started && session.state != .stopped && waitedMs < 20_000 {
            try? await Task.sleep(for: .milliseconds(100))
            waitedMs += 100
        }
        guard session.state == .started else {
            fail("Timed out connecting to the glasses. Are they on, unfolded, and paired in Meta AI?")
            return
        }

        let config = StreamConfiguration(videoCodec: .raw, resolution: .low, frameRate: 15)
        let camera: MWDATCamera.Camera?
        do {
            camera = try session.addCamera(config: config)
        } catch {
            fail("Could not open the glasses camera: \(error.localizedDescription)")
            return
        }
        guard let camera else {
            fail("The glasses session was not ready for a camera.")
            return
        }
        self.camera = camera
        let stream = camera.stream

        stream.statePublisher.listen { [weak self] state in
            guard let self else { return }
            switch state {
            case .streaming: self.onState?(.streaming)
            case .starting, .waitingForDevice: self.onState?(.waitingForDevice)
            case .paused: self.onState?(.paused)
            case .stopped, .stopping: if !self.failed { self.onState?(.idle) }
            }
        }.store(in: tokens)
        stream.errorPublisher.listen { error in
            NSLog("[Metanav] glasses stream error: \(error.description)")
        }.store(in: tokens)
        stream.videoFramePublisher.listen { [weak self] frame in
            guard let self, let onFrame = self.onFrame else { return }
            let buffer = frame.sampleBuffer
            if let pixelBuffer = CMSampleBufferGetImageBuffer(buffer) {
                onFrame(CapturedFrame(image: .pixelBuffer(pixelBuffer), timestampMs: nowMs()))
            } else if let cgImage = frame.makeUIImage()?.cgImage {
                // Compressed frame (should not happen with `.raw`), decoded by the SDK.
                onFrame(CapturedFrame(image: .cgImage(cgImage), timestampMs: nowMs()))
            }
        }.store(in: tokens)

        stream.start()
    }

    func stop() {
        tokens.clear()
        camera?.stop()
        camera = nil
        session?.stop()
        session = nil
        if !failed { onState?(.idle) }
    }

    private func fail(_ message: String) {
        NSLog("[Metanav] \(message)")
        failed = true
        stop()
        onState?(.error(message))
    }
}
